### PR TITLE
Create route for reading Learning Object revisions

### PR DESCRIPTION
**Pull request by nvisal1**
***Purpose***
The purpose of this PR is to create a route for retrieving Learning Object revisions. The route supports a `summary` query parameter.

***Story***
This PR is in response to story https://app.clubhouse.io/clarkcan/story/788/add-summary-query-parameter-to-revision-resource

***Implementation***
The new route is structured as
`GET /users/:username/learning-objects/:id/revisions/:id?summary`

Once the above route is hit, the following steps occur: 
1. Validate request by checking that the given `userId` and `learningObjectId` produce a Learning Object
2. Ensure that the requester has Learning Object read access by checking for author, admin, or editor privilege. 
3. Check that the given Learning Object has a revision.
4. Return full Learning Object or Learning Object Summary based on the value of the `summary` query parameter.

A new `UserGateway` interface was created in `/LearningObjects`. This interface is implemented by the `ModuleUserGateway` and `StubUserGateway` classes. The `ModuleUserGateway` class makes a request to `User Service` to retrieve a full `User` object for a given username. This is done when structuring a full Learning Object response.

***NOTE***
Review https://github.com/Cyber4All/learning-object-service/pull/413 first

New environment variable: `USER_API`

### CODE REVIEW

**Comment by gshaw1997** [comment]
@nvisal1 I switched the base temporarily to reduce review load since this branch branches off of https://github.com/Cyber4All/learning-object-service/tree/nvisal1/ch787/create-a-route-for-making-new-revisions.

Be sure to switch base back to `master` when #413 is merged.

File: src/LearningObjects/LearningObjectInteractor.ts [hunk 1]
```diff
@@ -1398,6 +1398,82 @@ export async function createLearningObjectRevision(params: {
...
+  summary?: boolean,
+}): Promise<LearningObject | LearningObjectSummary> {
+  const learningObject = await validateRequest({
+    userId: params.userId,
+    learningObjectId: params.learningObjectId,
+    dataStore: params.dataStore,
+  });
+
+  if (
+    learningObject.author.username === params.requester.username ||
```
Comment (gshaw1997) [inline src/LearningObjects/LearningObjectInteractor.ts]:
```suggestion
   requesterIsAuthor({requester, authorUsername: learningObject.author.username})
```

File: src/LearningObjects/LearningObjectInteractor.ts [hunk 2]
```diff
@@ -1423,13 +1499,14 @@ function sanitizeUpdates(
  * with the given userId and Learning Object Id.
  * If it does not find a Learning Object that matches
  * the given criteria, it throws a Resource Error.
+ * Otherwise, it returns a working copy Learning Object.
  * @param params
  */
 async function validateRequest(params: {
```
Comment (gshaw1997) [inline src/LearningObjects/LearningObjectInteractor.ts]: This function seems like it's doing more that the name suggests. May need to update

File: src/LearningObjects/adapters/LearningObjectRouteHandler.ts [hunk 3]
```diff
@@ -236,4 +253,5 @@ export function initializePrivate({
     getLearningObjectChildren,
   );
   router.post('/users/:userId/learning-objects/:learningObjectId/revisions', createRevision);
+  router.get('/users/:userId/learning-objects/:id/revisions/:id', getRevision);
```
Comment (gshaw1997) [inline src/LearningObjects/adapters/LearningObjectRouteHandler.ts]: Are we grabbing the `username` or `userId`? If `username` we should update route param here. If `userId`, why the switch? It seems to be inconsistent with the rest of our route structures

File: src/LearningObjects/LearningObjectInteractor.ts [hunk 4]
```diff
@@ -1398,6 +1398,82 @@ export async function createLearningObjectRevision(params: {
...
+
+    if (params.summary) {
+      return getLearningObjectRevisionSummary({
+        dataStore: params.dataStore,
+        requester: params.requester,
+        id: params.learningObjectId,
+        revision: learningObject.revision,
+      });
+    }
+    return getLearningObjectById({
```
Comment (gshaw1997) [inline src/LearningObjects/LearningObjectInteractor.ts]: How do we know that we are getting the correct revision here?

Comment (nvisal1) [reply src/LearningObjects/LearningObjectInteractor.ts]: I thought that only one revision could exist at a time. Is that not the case?

File: src/LearningObjects/LearningObjectInteractor.ts [hunk 5]
```diff
@@ -651,46 +654,72 @@ export async function getActiveLearningObjectSummary({
...
+  username: string;
+  summary?: boolean,
+}): Promise<LearningObject | LearningObjectSummary> {
   try {
-    const object = await dataStore.fetchLearningObjectRevisionSummary({
-      id,
-      revision,
+    if (revisionId === 0) {
+      throw new ResourceError(
+        `Cannot find revision for Learning Object ${learningObjectId}`,
```
Comment (gshaw1997) [inline src/LearningObjects/LearningObjectInteractor.ts]:
```suggestion
        `Cannot find revision ${revisionId} for Learning Object ${learningObjectId}`,
```

File: src/LearningObjects/adapters/LearningObjectRouteHandler.ts [hunk 6]
```diff
@@ -245,6 +246,29 @@ export function initializePrivate({
...
+        requester: req.user,
+        summary: toBoolean(req.query.summary),
+      };
+      const learningObjectRevision = await LearningObjectInteractor.getLearningObjectRevision(params);
+      if (params.summary) {
+        res.status(200).json(learningObjectRevision);
+      } else {
+        const object = learningObjectRevision as LearningObject;
+        res.status(200).json(object.toPlainObject());
+      }
```
Comment (DonSeannelly) [inline src/LearningObjects/adapters/LearningObjectRouteHandler.ts]: From our convo, we can shorten this down to this:
```
if (learningObjectRevision instanceof LearningObject) {
  learningObjectRevision = learningObjectRevision.toPlainObject();
}
res.status(200).json(learningObjectRevision);
```

File: src/LearningObjects/adapters/LearningObjectRouteHandler.ts [hunk 7]
```diff
@@ -245,6 +246,28 @@ export function initializePrivate({
...
+    try {
+      const params = {
+        ...req.params,
+        revisionId: parseInt(req.params.revisionId, 10),
+        dataStore,
+        library,
+        requester: req.user,
+        summary: toBoolean(req.query.summary),
+      };
+      let learningObjectRevision: LearningObject | LearningObjectSummary | Partial<LearningObject>;
```
Comment (DonSeannelly) [inline src/LearningObjects/adapters/LearningObjectRouteHandler.ts]: When would this be a Partial? I thought it was either a `LearningObject` or a `LearningObjectSummary`?

**Pull request merged**